cozo-core/src/storage/newrocks.rs

```diff
@@ -0,0 +1,535 @@
+    }
+
+    fn range_count<'a>(&'a self, lower: &[u8], upper: &[u8]) -> Result<usize>
+    where
+        's: 'a,
+    {
+        let db_tx = self
+            .db_tx
+            .as_ref()
+            .ok_or(miette!("Transaction already committed"))?;
+        let iter = db_tx.iterator(rocksdb::IteratorMode::From(
+            lower,
+            rocksdb::Direction::Forward,
+        ));
+        let count = iter
+            .take_while(|item| match item {
+                Ok((k, _)) => k.as_ref() < upper,
+                Err(_) => false,
+            })
+            .count();
+        Ok(count)
+    }
+
+
+    fn total_scan<'a>(&'a self) -> Box<dyn Iterator<Item = Result<(Vec<u8>, Vec<u8>)>> + 'a>
+    where
+        's: 'a,
+    {
+        match self.db_tx {
+            Some(ref db_tx) => {
+                Box::new(db_tx.iterator(rocksdb::IteratorMode::Start).map(|item| {
+                    item.map(|(k, v)| (k.to_vec(), v.to_vec()))
+                        .into_diagnostic()
+                        .wrap_err_with(|| "Error during total scan")
+                }))
+            }
+            None => Box::new(std::iter::once(Err(miette!("Transaction already committed")))),
+        }
+    }
+}
+
+pub(crate) struct NewRocksDbIterator<'a> {
+    inner: rocksdb::DBIteratorWithThreadMode<'a, rocksdb::Transaction<'a, OptimisticTransactionDB>>,
+    upper_bound: Vec<u8>,
+}
+
+impl<'a> Iterator for NewRocksDbIterator<'a> {
+    type Item = Result<Tuple>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(result) = self.inner.next() {
+            match result {
+                Ok((k, v)) => {
+                    if k.as_ref() >= self.upper_bound.as_slice() {
+                        return None;
+                    }
+                    return Some(Ok(decode_tuple_from_kv(&k, &v, None)));
+                }
+                Err(e) => return Some(Err(miette!("Iterator error: {}", e))),
+            }
+        }
+        None
+    }
+}
+
+pub(crate) struct NewRocksDbSkipIterator<'a> {
+    inner: rocksdb::DBIteratorWithThreadMode<'a, rocksdb::Transaction<'a, OptimisticTransactionDB>>,
+    upper_bound: Vec<u8>,
+    valid_at: ValidityTs,
+}
+
+impl<'a> Iterator for NewRocksDbSkipIterator<'a> {
+    type Item = Result<Tuple>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for result in &mut self.inner {
+            match result {
+                Ok((k, v)) => {
+                    if k.as_ref() >= self.upper_bound.as_slice() {
+                        return None;
+                    }
+                    if let Some(mut tup) =
+                        check_key_for_validity(&k, self.valid_at, None).0
+                    {
+                        extend_tuple_from_v(&mut tup, &v);
+                        return Some(Ok(tup));
+                    }
+                }
+                Err(e) => return Some(Err(miette!("Iterator error: {}", e))),
+            }
+        }
+        None
+    }
```

Comment:
```suggestion
    fn next(&mut self) -> Option<Self::Item> {
        loop {
            self.inner.set_mode(rocksdb::IteratorMode::From(&self.next_bound, rocksdb::Direction::Forward));
            match self.inner.next() {
                None => return None,
                Some(Ok((k_slice, v_slice))) => {
                    if self.upper_bound.as_slice() <= k_slice.as_ref() {
                        return None;
                    }

                    let (ret, nxt_bound) = check_key_for_validity(k_slice.as_ref(), self.valid_at, None);
                    self.next_bound = nxt_bound;
                    if let Some(mut tup) = ret {
                        extend_tuple_from_v(&mut tup, v_slice.as_ref());
                        return Some(Ok(tup));
                    }
                },
                Some(Err(e)) => return Some(Err(miette!("Iterator Error: {}", e))),
            }
        }
    }
```
Modeled off the implementation in rocks.rs, this solves your time travel test issues.